### PR TITLE
Time format

### PR DESCRIPTION
**Pull request by zisra**

### CODE REVIEW

**Comment by zisra** [comment]
Screenshot 2023-04-11 at 4 19 32 PM
Looks like this

File: src/video/state/types.ts [hunk 1]
```diff
@@ -30,6 +30,7 @@ export type VideoPlayerState = {
     isFocused: boolean; // is the video player the users focus? (shortcuts only works when its focused)
     leftControlHovering: boolean; // is the cursor hovered over the left side of player controls
     popoutBounds: null | DOMRect; // bounding box of current popout
+    timeFormat: 0 | 1 | 2; // Time format of the video player
   };
```
Comment (mrjvs) [inline src/video/state/types.ts:34]: use an enum for these possible, never use magic numbers.

File: src/video/state/providers/videoStateProvider.ts [hunk 2]
```diff
@@ -133,6 +133,10 @@ export function createVideoStateProvider(
       // update localstorage
       setStoredVolume(volume);
     },
+    setTimeFormat(num) {
+      state.interface.timeFormat = num;
+      updateInterface(descriptor, state);
+    },
```
Comment (mrjvs) [inline src/video/state/providers/videoStateProvider.ts:139]: add this method to the `controls.ts` file, not to the state providers (as the format of the time has nothing to do with how the video is played)

File: src/video/state/logic/interface.ts [hunk 3]
```diff
@@ -18,6 +20,10 @@ function getInterfaceFromState(state: VideoPlayerState): VideoInterfaceEvent {
     isFocused: state.interface.isFocused,
     isFullscreen: state.interface.isFullscreen,
     popoutBounds: state.interface.popoutBounds,
+    timeFormat: state.interface.timeFormat,
+    setTimeFormat(timeFormat: 0 | 1 | 2) {
+      state.stateProvider?.setTimeFormat(timeFormat);
+    },
```
Comment (mrjvs) [inline src/video/state/logic/interface.ts:26]: dont add methods to state only objects, when you've added to the controls.ts file, you can do useControls to access the method.

File: src/video/components/actions/TimeAction.tsx [hunk 4]
```diff
@@ -45,11 +49,51 @@ export function TimeAction(props: Props) {
...
+      }}
+    >
+      <div
+        className={[
+          "flex items-center justify-center rounded-full bg-denim-600 bg-opacity-0 p-2 transition-colors duration-100 group-hover:bg-opacity-50 group-active:bg-denim-500 group-active:bg-opacity-100 sm:px-4",
+        ].join(" ")}
+      >
+        <div className={props.className}>
+          <p className="select-none text-white">
+            {/* {time} {props.noDuration ? "" : `/ ${duration}`} */}
```
Comment (mrjvs) [inline src/video/components/actions/TimeAction.tsx:82]: please remove the comment

File: src/video/components/actions/TimeAction.tsx [hunk 5]
```diff
@@ -45,11 +49,51 @@ export function TimeAction(props: Props) {
...
-    <div className={props.className}>
-      <p className="select-none text-white">
-        {time} {props.noDuration ? "" : `/ ${duration}`}
-      </p>
-    </div>
+    <button
+      type="button"
+      className={[
+        "group pointer-events-auto translate-x-[-16px] text-white transition-transform duration-100 active:scale-110",
+      ].join(" ")}
```
Comment (mrjvs) [inline src/video/components/actions/TimeAction.tsx:70]: dont offset the by -16 pixels, just remove it

File: src/video/components/actions/TimeAction.tsx [hunk 6]
```diff
@@ -45,11 +49,51 @@ export function TimeAction(props: Props) {
...
+              ? `${time} ${props.noDuration ? "" : `/ ${duration}`}`
+              : `${t("videoPlayer.timeLeft", {
+                  timeLeft,
+                })}${
+                  videoTime.time === videoTime.duration
+                    ? ""
+                    : ` - ${t("videoPlayer.finishAt", {
+                        timeFinished,
+                      })}`
+                } `}
```
Comment (mrjvs) [inline src/video/components/actions/TimeAction.tsx:93]: for mobile (`useIsMobile()`). only display the time left with negative sign in front for the second time format.
Also this conditional may be worth putting in a variable with normal if statements, its getting quite hard to read

File: src/video/components/actions/TimeAction.tsx [hunk 7]
```diff
@@ -45,11 +49,51 @@ export function TimeAction(props: Props) {
   );
   const duration = formatSeconds(videoTime.duration, hasHours);
 
+  const timeLeft = formatSeconds(
+    (videoTime.duration - videoTime.time) / mediaPlaying.playbackSpeed
+  );
```
Comment (mrjvs) [inline src/video/components/actions/TimeAction.tsx:54]: don't forget about the hasHours parameter, right now it doesn't show hours in time left

**Review by mrjvs — CHANGES_REQUESTED** [review]

File: src/video/components/actions/TimeAction.tsx (hunk 5 above)
Comment (zisra) [reply src/video/components/actions/TimeAction.tsx:70]: The 16px makes it look exactly the way it looked before my changes, and If I were to remove it the time would be way far too the right

Comment (mrjvs) [reply src/video/components/actions/TimeAction.tsx:70]: it looks wonky on mobile, so I think its best to remove it

File: src/video/state/logic/interface.ts (hunk 3 above)
Comment (zisra) [reply src/video/state/logic/interface.ts]: "React Hook "useControls" cannot be called at the top level. React Hooks must be called in a React function component or a custom React Hook function."

**Comment by zisra** [comment]
Not sure whether this fixes it or not. Sorry if I did it wrong

**Review by mrjvs — APPROVED** [review]

**Pull request merged**